tests/console/nvme_checks.pm

```diff
@@ -0,0 +1,116 @@
+
+use base "installbasetest";
+use strict;
+use warnings;
+use testapi;
+use utils;
+use scheduler 'get_test_suite_data';
+
+sub _smoke_checks {
+    my $nvm_test_data = shift;
+    assert_script_run("lspci -nn | grep -i nvm");
+    assert_script_run("test -d /sys/class/nvme");
+    assert_script_run("test -c /dev/nvme || echo 'nvm character devices not found'");
+
+    my @nvm_partitions = split(/\n/, script_output("ls -l /dev/$nvm_test_data->{nvm_disk}*"));
+    foreach (@nvm_partitions) {
+        assert_script_run("test -b $_ || echo 'nvm is not a block device'");
+    }
+}
+
+sub _exercise_nvme_commands_and_validate_output {
+    my $nvm_test_data = shift;
+    # send Identify Admin Command to the NVMe controller.
+    assert_script_run("nvme id-ctrl -H /dev/$nvm_test_data->{nvm_char_device}");
+    save_screenshot;
+    my $MODE       = '\/dev\/' . $nvm_test_data->{nvm_char_device};
+    my $SN         = $nvm_test_data->{sn};
+    my $MODEL      = $nvm_test_data->{model};
+    my $NAMESPACE  = $nvm_test_data->{namespace_count};
+    my $DISK_SIZE  = $nvm_test_data->{nvm_disk_size};
+    my $BLOCK_SIZE = $nvm_test_data->{nvm_block_size};
+    my $list_regex = qr/${MODE}\s+$SN\s+${MODEL}\s+${NAMESPACE}\s+${DISK_SIZE}\s+GB.*\d+\s+GB\s+${BLOCK_SIZE}/;
+    validate_script_output("nvme list",                                         sub { $list_regex });
+    validate_script_output("nvme list-ns /dev/nvme0 | awk -F ':' '{print $2}'", sub { /0x1/ });
+    # the details of this namespace
+    # check LBA Format fields
+    # ms = Metadata Size, 0 means metadata is not supported
+    # lbads = LBA Data Size, in terms of a power of two, so LBA data size is 2⁹ = 512 bytes
+    # rp = Relative Performance, the lower the better
+    validate_script_output("nvme id-ns /dev/nvme0n1 --namespace-id=0x1", sub { m/lbaf\s+0\s+:\s+ms:0\s+lbads:9\s+rp:0\s\(in use\)/ });
+    # show registers
+    # https://bugzilla.suse.com/show_bug.cgi?id=1175217
+    my $ret_code = script_run("nvme show-regs -H /dev/nvme0n1");
+    record_soft_failure "bsc#1175217 - not permitted shared memory access to get registers" if ($ret_code > 0);
+}
+
+sub _issue_read_command_and_compare_with_dd {
+    my $args = shift;
+    assert_script_run("dd if=/dev/nvme0n1 of=lba.0.dd bs=512 count=1");
+    # read LBA 0 through nvme utility
+    assert_script_run("nvme read /dev/nvme0n1 --start-block=0 --block-count=0 --data-size=512 --data=lba.0.read");
+    # Read is opcode 0x02, we are sending command to namespace 0x1, reading 512 bytes
+    # code word 10 and 11 (cdw10 and cdw11) specifies the start of reading in LBA blocks,
+    # and the Bits 15:00 of code word 12 (cdw12) is number of blocks to read (set to 0 to indicate 1 block will be read).
+    assert_script_run("nvme io-passthru /dev/nvme0 --opcode=0x02 --namespace-id=0x1 --data-len=512 --read --cdw10=0 --cdw11=0 --cdw12=0 -b > lba.0.io");
+    assert_script_run("cmp lba.0.dd lba.0.read");
```

Comment:
a message error here would be helpful for further investigation of those two comparisons.

Reply:
I havent checked in the code above but the module should fail if cmp differs. When the `cmp` fails provides a error like `lba.0.dd lba.0.read differ: byte 1, line 1`. do you think it would be more descriptive if there was a custom message? And in additional i think the name of subroutine show the intention and the expectations. wdyt?

Reply:
It is not clear from the code why you are comparing in two different ways (I have to google it) so a descriptive error message or a simple comment why those two comparison are needed should be enough.

Reply:
it is a small validation of the `read` operation. The result should not be differ between the NVMe and from the manual execution. From my understanding, this test is needed because there is a separate pair of submission and completion queue for Admin Commands. The other is IO cmds. We validate consistency. again not an expert here

Reply:
I read the same that you read I think :) so if we can help the next reader not included in this conversation to understand, error message or comment, would be nice.

Reply:
fail_message added

Reply:
fail_message added